### ISSUE TITLE
--disallow-any=decorated error for classmethod with Any in __init__ argument

### ISSUE THREAD

**Issue by JelleZijlstra** [python/mypy#3809]
Running `mypy  --disallow-any=decorated disallow.py` (with mypy master) on
```python
from typing import Any

class B:
    def __init__(self, *args: Any) -> None:
        pass

    @classmethod
    def clsmethod(cls) -> int:
        return 0
```
produces
```
disallow.py:9: error: Type of decorated function contains type "Any" (Callable[[Type[B]], int])
```
This is unintuitive because no `Any` appears in the displayed type; in my actual code the `__init__` was far away from the class definition.

I'm open to persuasion otherwise, but I think we shouldn't generate an error here; we should just generate a `--disallow-any` error for the constructor.

**Comment by gvanrossum** [comment]
I agree this is a bug in that flag.

**Comment by ilinum** [comment]
Hmm, that's interesting. It definitely does seem like a bug.

So the reason why this is reported as an error is that actual type of `cls` in that case is `Callable[[*Any], a.B]`. It is a so-called `type_object_type` ([here](https://github.com/python/mypy/pull/3718#issuecomment-317047231) is the explanation of what `type_object_type` is).

Note that if #3718 was merged, this error would go away, since `--disallow-any=expr` and `--disallow-any=decorated` share code for detecting Anys.

Personally, I think #3718 was the right fix for both `--disallow-any=expr` and `--disallow-any=decorated`. But David seemed very adamant that we should report such errors for `--disallow-any=expr`.

It would be pretty trivial to fix it just for `--disallow-any=decorated` (and not for `--disallow-any=expr`). Should I go ahead and do that?

**Comment by ilevkivskyi** [comment]
FWIW this is fixed on master.
